Source/CarthageKit/Project.swift

```diff
@@ -540,38 +540,57 @@ public final class Project {
 					.map { (cartfile: Cartfile) -> DependencyGraph in
 						return [ dependency.project: Set(cartfile.dependencies.map { $0.project }) ]
 					}
+					.concat(SignalProducer(value: [ dependency.project: Set() ]))
+					.take(1)
 			}
-			.collect()
-			.flatMap(.Latest) { (graphs: [DependencyGraph]) -> SignalProducer<Dependency<PinnedVersion>, CarthageError> in
-				let graph = graphs.reduce([:]) { (graph: DependencyGraph, item: DependencyGraph) -> DependencyGraph in
-					var result = graph
-					for (dependency, projects) in item {
-						var recursive = projects
-						for project in projects {
-							if let projects = result[project] {
-								recursive.unionInPlace(projects)
-							}
-						}
-						
-						result.updateValue(recursive, forKey: dependency)
-						
-						for (project, projects) in result {
-							if projects.contains(dependency) {
-								result.updateValue(projects.union(recursive), forKey: project)
-							}
+			.reduce([:]) { (working: DependencyGraph, next: DependencyGraph) in
+				var result = working
+				next.forEach { result.updateValue($1, forKey: $0) }
+				return result
+			}
+			.flatMap(.Latest) { (graph: DependencyGraph) -> SignalProducer<Dependency<PinnedVersion>, CarthageError> in
+				// Open Question: Is there a place that a generic topological 
+				// sort should live?
+				func topologicalSort<Node: Equatable>(edges: Dictionary<Node, Set<Node>>, @noescape isOrderedBefore: (Node, Node) -> Bool) -> [Node] {
+					var nodeQueue: [Node] = edges.filter { $1.isEmpty }.map { return $0.0 }
```

Comment:
I'd prefer to use named variables in the blocks. I think that makes it easier to read the code. It will probably make sense to split the `filter` and `map` on separate lines with the added characters.